test/embedding/embeddingdl.c

```diff
@@ -0,0 +1,198 @@
+#endif
+{
+#ifdef _OS_WINDOWS_
+    void *p = GetProcAddress(libjulia, name);
+#else
+    void *p = dlsym(libjulia, name);
+#endif
+
+    // Unfortunately Julia renames jl_init to jl_init__threading if
+    // Julia is compiled with threading support, so we have to check
+    // which of these is available, or otherwise query Julia in some
+    // other way (issue #28824).
+    if (!p && strcmp(name, "jl_init") == 0) {
+#ifdef _OS_WINDOWS_
+        p = GetProcAddress(libjulia, "jl_init__threading");
+#else
+        p = dlsym(libjulia, "jl_init__threading");
+#endif
+    }
+
+    if (!p) {
+        fprintf(stderr, "%s not found in libjulia.\n", name);
+        *success = 0;
+    }
+
+    return p;
+}
+
+// Open libjulia and extract pointers to the needed functions.
+static int load_julia_functions()
+{
+#ifdef _OS_WINDOWS_
+    // libjulia.dll needs to be in the same directory as the
+    // executable or in PATH.
+    const char *library_name = "libjulia.dll";
+    HMODULE libjulia = LoadLibrary(library_name);
+#else
+    // libjulia.so needs to be in LD_LIBRARY_PATH. It could also be in
+    // rpath (but that kind of defeats the purpose of dynamically
+    // loading libjulia) or an absolute path could be given, computed
+    // from other information.
+#ifdef __APPLE__
+    const char *library_name = "libjulia.dylib";
+#else
+    const char *library_name = "libjulia.so";
+#endif
+    void *libjulia = dlopen(library_name, RTLD_LAZY | RTLD_GLOBAL);
+#endif
+
+    if (!libjulia) {
+        fprintf(stderr, "Failed to load libjulia.\n");
+        return 0;
+    }
+
+    int success = 1;
+    p_jl_init = load_function(libjulia, "jl_init", &success);
+    p_jl_is_initialized = load_function(libjulia, "jl_is_initialized", &success);
+    p_jl_atexit_hook = load_function(libjulia, "jl_atexit_hook", &success);
+    p_jl_eval_string = load_function(libjulia, "jl_eval_string", &success);
+    p_jl_symbol = load_function(libjulia, "jl_symbol", &success);
+    p_jl_main_module = load_function(libjulia, "jl_main_module", &success);
+    p_jl_get_global = load_function(libjulia, "jl_get_global", &success);
+    p_jl_unbox_voidpointer = load_function(libjulia, "jl_unbox_voidpointer", &success);
+    p_jl_string_ptr = load_function(libjulia, "jl_string_ptr", &success);
+    p_jl_exception_occurred = load_function(libjulia, "jl_exception_occurred", &success);
+
+    return success;
+}
+
+// Masquerade the dynamically loaded function pointers as regular functions.
+#define jl_init (*p_jl_init)
+#define jl_is_initialized (*p_jl_is_initialized)
+#define jl_atexit_hook (*p_jl_atexit_hook)
+#define jl_eval_string (*p_jl_eval_string)
+#define jl_symbol (*p_jl_symbol)
+#define jl_main_module (*p_jl_main_module)
+#define jl_get_global (*p_jl_get_global)
+#define jl_unbox_voidpointer (*p_jl_unbox_voidpointer)
+#define jl_string_ptr (*p_jl_string_ptr)
+#define jl_exception_occurred (*p_jl_exception_occurred)
```

Comment:
Any particular advantage to having it in `julia.h` rather than in a separate file? I actually found it helpful not having to include `julia.h` as that gave me conflicts with some qt header file.

Reply:
Only that `julia.h` is currently the place where the public interface is defined.

The idea of defining the symbol `JL_DLOPEN_INTERFACE` is that it would strip julia.h down to almost nothing so you shouldn't have any conflicts with other libraries such as Qt. Do you remember what the conflicts were?

Reply:
No, that was too long ago now.